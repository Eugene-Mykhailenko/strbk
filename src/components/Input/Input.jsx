/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/no-access-state-in-setstate */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { CSSTransition } from 'react-transition-group';
import classNames from 'classnames';
import styles from './Input.scss';
import animationStyles from '../../assets/styles/base/animations.scss';
import Icon from '../Icon/Icon';

class Input extends Component {
  constructor(props) {
    super(props);

    // TODO: decide if we need dynamic type update.
    this.state = {
      type: props.type,
      isPasswordType: props.type === 'password',
      isPasswordVisible: false,
      hasFocus: false,
    };

    this.internalOnChange = this.internalOnChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.togglePassword = this.togglePassword.bind(this);
  }

  onFocus() {
    this.setState({
      hasFocus: true,
    });
  }

  onBlur(event) {
    if (!event.target.value.length) {
      this.setState({
        hasFocus: false,
      });
    }
  }

  getContainerClassNames() {
    const { hasError, isRequired, disabled } = this.props;
    const { hasFocus, isPasswordType } = this.state;

    return classNames(styles['form-group'], {
      [styles['has-focus']]: hasFocus,
      [styles['is-required']]: isRequired,
      [styles['is-password-type']]: isPasswordType,
      [styles['has-error']]: hasError,
      [styles['is-disabled']]: disabled,
    });
  }

  getInputClassNames() {
    const { hasError } = this.props;
    const { hasFocus } = this.state;

    return classNames(styles['form-control'], {
      [styles['has-focus']]: hasFocus,
      [styles['has-error']]: hasError,
    });
  }

  togglePassword() {
    const isPasswordVisible = !this.state.isPasswordVisible;

    this.setState({
      isPasswordVisible,
      type: isPasswordVisible ? 'text' : 'password',
    });
  }

  internalOnChange(event) {
    const { onChange, disabled } = this.props;

    if (!disabled) {
      onChange(event);
    }
  }

  renderLabel() {
    const { placeholder } = this.props;

    return placeholder && <label className={styles.label}>{placeholder}</label>;
  }

  renderButton() {
    const { isPasswordVisible, isPasswordType } = this.state;

    return (
      isPasswordType && (
        <button className={styles.btn} onClick={this.togglePassword}>
          <Icon
            className={styles.icon}
            width={24}
            height={24}
            glyph={isPasswordVisible ? 'UII_Eye' : 'UII_EyeSlash'}
          />
        </button>
      )
    );
  }

  renderErrorText() {
    const { hasError, errorText } = this.props;

    return (
      <CSSTransition
        in={hasError}
        classNames={{
          enter: animationStyles['fade-in-start'],
          enterActive: animationStyles['fade-in-end'],
          exit: animationStyles['fade-out-start'],
          exitActive: animationStyles['fade-out-end'],
        }}
        timeout={300}
        unmountOnExit
      >
        <div className={styles['error-text']}>{errorText}</div>
      </CSSTransition>
    );
  }

  render() {
    const { disabled } = this.props;
    const { type } = this.state;

    return (
      <div className={this.getContainerClassNames()}>
        {this.renderLabel()}
        <input
          className={this.getInputClassNames()}
          type={type}
          disabled={disabled}
          onChange={this.internalOnChange}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
        />
        {this.renderButton()}
        {this.renderErrorText()}
      </div>
    );
  }
}

Input.propTypes = {
  /** Input Placeholder/Label */
  placeholder: PropTypes.string,
  /** Input field type: text, number, password and etc */
  type: PropTypes.string,
  /** Callback function for event */
  onChange: PropTypes.func.isRequired,
  /** Add star for label if this field is required */
  isRequired: PropTypes.bool,
  /** Error indication */
  hasError: PropTypes.bool,
  /** Error message */
  errorText: PropTypes.string,
  /** Disabled input */
  disabled: PropTypes.bool,
};

Input.defaultProps = {
  placeholder: undefined,
  type: 'text',
  isRequired: false,
  hasError: false,
  errorText: undefined,
  disabled: false,
};

export default Input;
