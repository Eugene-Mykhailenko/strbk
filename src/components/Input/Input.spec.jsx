import React from 'react';
import Input from './Input';

jest.mock('../Icon/importIcons');

const config = {
  type: 'text',
  placeholder: 'Label',
  onChange: jest.fn(),
  isRequired: true,
  hasError: true,
  errorText: 'Error Text exemple',
  disabled: true,
};

describe('Input', () => {
  it('should render input of text type by default', () => {
    const input = shallow(<Input onChange={config.onChange} />);

    expect(input.find('input').prop('type')).toEqual('text');
    expect(input).toMatchSnapshot();
  });

  it('should render floating label', () => {
    const input = shallow(
      <Input placeholder={config.placeholder} onChange={config.onChange} />,
    );

    expect(input.find('label').text()).toEqual(config.placeholder);
    expect(input).toMatchSnapshot();
  });

  it('should be able to render required state', () => {
    const input = shallow(
      <Input isRequired={config.isRequired} onChange={config.onChange} />,
    );

    expect(input.prop('className')).toMatch(/is-required/);
    expect(input).toMatchSnapshot();
  });

  it('should be able to render error state', () => {
    const input = shallow(
      <Input hasError={config.hasError} onChange={config.onChange} />,
    );

    expect(input.prop('className')).toMatch(/has-error/);
    expect(input.find('input').prop('className')).toMatch(/has-error/);
    expect(input).toMatchSnapshot();
  });

  it('should be able to render error message', () => {
    const input = shallow(
      <Input
        hasError={config.hasError}
        errorText={config.errorText}
        onChange={config.onChange}
      />,
    );

    expect(input.prop('className')).toMatch(/has-error/);
    expect(input.find('input').prop('className')).toMatch(/has-error/);
    expect(input.find('.error-text').text()).toEqual(config.errorText);
    expect(input).toMatchSnapshot();
  });

  it('should be able to render disabled state', () => {
    const input = shallow(
      <Input disabled={config.disabled} onChange={config.onChange} />,
    );

    expect(input.prop('className')).toMatch(/is-disabled/);
    expect(input.find('input').prop('disabled')).toEqual(true);
    expect(input).toMatchSnapshot();
  });

  it('should not trigger onChange handler in a disabled state', () => {
    const input = shallow(
      <Input disabled={config.disabled} onChange={config.onChange} />,
    );

    input.find('input').simulate('change');

    expect(config.onChange).toHaveBeenCalledTimes(0);
    expect(input).toMatchSnapshot();
  });

  it('should trigger onChange handler', () => {
    const input = shallow(<Input onChange={config.onChange} />);

    input.find('input').simulate('change');

    expect(config.onChange).toHaveBeenCalledTimes(1);
    expect(input).toMatchSnapshot();
  });

  it('should be able to render input of password type', () => {
    const input = shallow(<Input type="password" onChange={config.onChange} />);

    expect(input.find('input').prop('type')).toEqual('password');
    expect(input).toMatchSnapshot();
  });

  it('should render toggle password button for input of password type', () => {
    const input = shallow(<Input type="password" onChange={config.onChange} />);

    expect(input.find('input').prop('type')).toEqual('password');
    expect(input.find('button').length).toEqual(1);
    expect(input).toMatchSnapshot();
  });

  it('should add correct styles on focus', () => {
    const input = shallow(<Input type="password" onChange={config.onChange} />);

    input.find('input').simulate('focus');

    expect(input.prop('className')).toMatch(/has-focus/);
    expect(input.find('input').prop('className')).toMatch(/has-focus/);
    expect(input).toMatchSnapshot();
  });

  it('should disabled focus event', () => {
    const input = shallow(<Input type="password" onChange={config.onChange} />);

    input.find('input').simulate('focus');

    expect(input.prop('className')).toMatch(/has-focus/);
    expect(input.find('input').prop('className')).toMatch(/has-focus/);

    input.find('input').simulate('blur', { target: { value: '' } });

    expect(input.prop('className')).not.toMatch(/has-focus/);
    expect(input.find('input').prop('className')).not.toMatch(/has-focus/);
    expect(input).toMatchSnapshot();
  });
});
