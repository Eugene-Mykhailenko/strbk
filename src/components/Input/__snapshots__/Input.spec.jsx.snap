// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should add correct styles on focus 1`] = `
<div
  className="form-group has-focus is-password-type"
>
  <input
    className="form-control has-focus"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="password"
  />
  <button
    className="btn"
    onClick={[Function]}
  >
    <Icon
      className="icon"
      fill="currentColor"
      glyph="UII_EyeSlash"
      height={24}
      width={24}
    />
  </button>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should be able to render disabled state 1`] = `
<div
  className="form-group is-disabled"
>
  <input
    className="form-control"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should be able to render error message 1`] = `
<div
  className="form-group has-error"
>
  <input
    className="form-control has-error"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={true}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    >
      Error Text exemple
    </div>
  </CSSTransition>
</div>
`;

exports[`Input should be able to render error state 1`] = `
<div
  className="form-group has-error"
>
  <input
    className="form-control has-error"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={true}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should be able to render input of password type 1`] = `
<div
  className="form-group is-password-type"
>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="password"
  />
  <button
    className="btn"
    onClick={[Function]}
  >
    <Icon
      className="icon"
      fill="currentColor"
      glyph="UII_EyeSlash"
      height={24}
      width={24}
    />
  </button>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should be able to render required state 1`] = `
<div
  className="form-group is-required"
>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should disabled focus event 1`] = `
<div
  className="form-group is-password-type"
>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="password"
  />
  <button
    className="btn"
    onClick={[Function]}
  >
    <Icon
      className="icon"
      fill="currentColor"
      glyph="UII_EyeSlash"
      height={24}
      width={24}
    />
  </button>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should not trigger onChange handler in a disabled state 1`] = `
<div
  className="form-group is-disabled"
>
  <input
    className="form-control"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should render floating label 1`] = `
<div
  className="form-group"
>
  <label
    className="label"
  >
    Label
  </label>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should render input of text type by default 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should render toggle password button for input of password type 1`] = `
<div
  className="form-group is-password-type"
>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="password"
  />
  <button
    className="btn"
    onClick={[Function]}
  >
    <Icon
      className="icon"
      fill="currentColor"
      glyph="UII_EyeSlash"
      height={24}
      width={24}
    />
  </button>
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;

exports[`Input should trigger onChange handler 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
  <CSSTransition
    classNames={
      Object {
        "enter": "fade-in-start",
        "enterActive": "fade-in-end",
        "exit": "fade-out-start",
        "exitActive": "fade-out-end",
      }
    }
    in={false}
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="error-text"
    />
  </CSSTransition>
</div>
`;
